File: services/csharp/GlobularClient/GlobularClient.cs
using System.Net.Http;
using System.Text.Json;
using System.IO;
using System.Threading.Tasks;
using Grpc.Core;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;

namespace Globular
{
    /** Globular server config. **/
    public class ServerConfig
    {
        public string Domain { get; set; }
        public string Name { get; set; }
        public string Protocol { get; set; }
        public string CertStableURL { get; set; }
        public string CertURL { get; set; }
        public int ConfigurationPort { get; set; }
        public int PortHttp { get; set; }
        public int PortHttps { get; set; }
        public int AdminPort { get; set; }
        public string AdminEmail { get; set; }
        public int RessourcePort { get; set; }
        public int ServicesDiscoveryPort { get; set; }
        public int ServicesRepositoryPort { get; set; }
        public int CertificateAuthorityPort { get; set; }
        public int LoadBalancingServicePort { get; set; }
        public int SessionTimeout { get; set; }
        public int CertExpirationDelay { get; set; }
        public int IdleTimeout { get; set; }

        public string[] Discoveries { get; set; }
        public string[] DNS { get; set; }

        // The map of service object.
        public Dictionary<string, ServiceConfig> Services { get; set; }

    }

    /**
     * Used by JSON serialysation.
     */
    public class ServiceConfig
    {
        public string CertAuthorityTrust { get; set; }
        public string CertFile { get; set; }
        public string KeyFile { get; set; }
        public string Domain { get; set; }
        public string Name { get; set; }
        public string Id { get; set; }
        public string Path { get; set; }
        public string Proto { get; set; }
        public int Port { get; set; }
        public bool TLS { get; set; }
    }

    public class Client
    {
    	private string id;
        private string name;
        private string domain;
        private int port;
        private bool hasTls;
        private string caFile;
        private string keyFile;
        private string certFile;
        
        protected Channel channel;

        // Get Domain return the client domain.
        public string GetDomain()
        {
            return this.domain;
        }
        
        public string GetId()
        {
            return this.id;
        }


        public string GetName()
        {
            return this.name;
        }

        public int GetPort()
        {
            return this.port;
        }

        // Close the client.
        public void Close()
        {
            // close the connection channel.
            this.channel.ShutdownAsync().Wait();
        }

        // At firt the port contain the http(s) domain of the globular server.
        // The configuration will be get from that domain and the port will
        // be set back to the correct domain.
        public void SetPort(int port)
        {
            this.port = port;
        }

        // Set the name of the client
        public void SetName(string name)
        {
            this.name = name;
        }

        // Set the domain of the client
        public void SetDomain(string domain)
        {
            this.domain = domain;
        }

        ////////////////// TLS ///////////////////

        /// <summary>
        /// Test if the server is secure with TLS.
        /// </summary>
        /// <returns>True if it secure.</returns>
        public bool HasTLS()
        {
            return this.hasTls;
        }

        // Get the TLS certificate file path
        public string GetCertFile()
        {
            return this.certFile;
        }

        // Get the TLS key file path
        public string GetKeyFile()
        {
            return this.keyFile;
        }

        // Get the TLS key file path
        public string GetCaFile()
        {
            return this.certFile;
        }

        // Set the client is a secure client.
        public void SetTLS(bool hasTls)
        {
            this.hasTls = hasTls;
        }

        // Set TLS certificate file path
        public void SetCertFile(string certFile)
        {
            this.certFile = certFile;
        }

        // Set TLS key file path
        public void SetKeyFile(string keyFile)
        {
            this.keyFile = keyFile;
        }

        // Set TLS authority trust certificate file path
        public void SetCaFile(string caFile)
        {
            this.caFile = caFile;
        }

        // Get the CA public certificate.
        private string getCaCertificate(string domain, int ConfigurationPort)
        {
            // Get the configuration from the globular server.
            var client = new HttpClient();
            string rqst = "http://" + domain + ":" + ConfigurationPort + "/get_ca_certificate";
            var task = Task.Run(() => client.GetAsync(rqst));
            task.Wait();
            var rsp = task.Result;
            if (rsp.IsSuccessStatusCode == false)
            {
                throw new System.InvalidOperationException("Fail to get client configuration " + rqst);
            }

            return rsp.Content.ReadAsStringAsync().Result;
        }

        public static string Base64Encode(string plainText)
        {
            var plainTextBytes = System.Text.Encoding.UTF8.GetBytes(plainText);
            return System.Convert.ToBase64String(plainTextBytes);
        }

        // ask globular CA to sign the cerificate.
        private string signCaCertificate(string domain, int ConfigurationPort, string csr)
        {
            var client = new HttpClient();

            string csr_str = Base64Encode(csr);
            string rqst = "http://" + domain + ":" + ConfigurationPort + "/sign_ca_certificate?csr=" + csr_str;
            var task = Task.Run(() => client.GetAsync(rqst));
            task.Wait();
            var rsp = task.Result;
            if (rsp.IsSuccessStatusCode == false)
            {
                throw new System.InvalidOperationException("Fail to get sign ca certificate!");
            }

            return rsp.Content.ReadAsStringAsync().Result;
        }

        /**
         * I will made use of openssl as external command to be able to generate key and
         * certificate the same way in every language.
         */
        private void generateClientPrivateKey(string path, string pwd)
        {
            if (File.Exists(path + "/client.key"))
            {
                return;
            }

            Process process_0 = new Process();
            process_0.StartInfo.FileName = "openssl.exe";

            // Set args
            process_0.StartInfo.ArgumentList.Add("genrsa");
            process_0.StartInfo.ArgumentList.Add("-passout");
            process_0.StartInfo.ArgumentList.Add("pass:" + pwd);
            process_0.StartInfo.ArgumentList.Add("-des3");
            process_0.StartInfo.ArgumentList.Add("-out");
            process_0.StartInfo.ArgumentList.Add(path + "/client.pass.key");
            process_0.StartInfo.ArgumentList.Add("4096");

            // set options
            process_0.StartInfo.UseShellExecute = false;
            process_0.StartInfo.RedirectStandardOutput = true;
            process_0.StartInfo.RedirectStandardError = true;

            process_0.Start();
            process_0.WaitForExit();

            Process process_1 = new Process();
            process_1.StartInfo.FileName = "openssl.exe";

            // Set args
            process_1.StartInfo.ArgumentList.Add("rsa");
            process_1.StartInfo.ArgumentList.Add("-passin");
            process_1.StartInfo.ArgumentList.Add("pass:" + pwd);
            process_1.StartInfo.ArgumentList.Add("-in");
            process_1.StartInfo.ArgumentList.Add(path + "/client.pass.key");
            process_1.StartInfo.ArgumentList.Add("-out");
            process_1.StartInfo.ArgumentList.Add(path + "/client.key");

            // set options
            process_1.StartInfo.UseShellExecute = false;
            process_1.StartInfo.RedirectStandardOutput = true;
            process_1.StartInfo.RedirectStandardError = true;
            process_1.Start();
            process_1.WaitForExit();

            // remove the intermediary file.
            File.Delete(path + "/client.pass.key");
        }

        /**
         * Generate a client signing request for a given domain.
         */
        private void generateClientCertificateSigningRequest(string path, string domain)
        {
            if (File.Exists(path + "/client.csr"))
            {
                return;
            }

            Process process_0 = new Process();
            process_0.StartInfo.FileName = "openssl.exe";

            // Set args
            process_0.StartInfo.ArgumentList.Add("req");
            process_0.StartInfo.ArgumentList.Add("-new");
            process_0.StartInfo.ArgumentList.Add("-key");
            process_0.StartInfo.ArgumentList.Add(path + "/client.key");
            process_0.StartInfo.ArgumentList.Add("-out");
            process_0.StartInfo.ArgumentList.Add(path + "/client.csr");
            process_0.StartInfo.ArgumentList.Add("-subj");
            process_0.StartInfo.ArgumentList.Add("/CN=" + domain);

            // set options
            process_0.StartInfo.UseShellExecute = false;
            process_0.StartInfo.RedirectStandardOutput = true;
            process_0.StartInfo.RedirectStandardError = true;

            process_0.Start();
            process_0.WaitForExit();
        }

        private void keyToPem(string name, string path, string pwd)
        {
            if (File.Exists(path + "/" + name + ".pem"))
            {
                return;
            }

            Process process_0 = new Process();
            process_0.StartInfo.FileName = "openssl.exe";

            // Set args
            process_0.StartInfo.ArgumentList.Add("pkcs8");
            process_0.StartInfo.ArgumentList.Add("-topk8");
            process_0.StartInfo.ArgumentList.Add("-nocrypt");
            process_0.StartInfo.ArgumentList.Add("-passin");
            process_0.StartInfo.ArgumentList.Add("pass:" + pwd);
            process_0.StartInfo.ArgumentList.Add("-in");
            process_0.StartInfo.ArgumentList.Add(path + "/" + name + ".key");
            process_0.StartInfo.ArgumentList.Add("-out");
            process_0.StartInfo.ArgumentList.Add(path + "/" + name + ".pem");

            // set options
            process_0.StartInfo.UseShellExecute = false;
            process_0.StartInfo.RedirectStandardOutput = true;
            process_0.StartInfo.RedirectStandardError = true;

            process_0.Start();
            process_0.WaitForExit();
        }

        private static bool VerifyPeer(VerifyPeerContext context)
        {
            File.WriteAllText("c:/temp/toto.txt", "VerifiPeer!");
            return true;
        }

        private void init(string id , string domain="localhost",  int configurationPort=10000)
        {
            // Get the configuration from the globular server.
            var client = new HttpClient();
            string rqst = "http://" + domain + ":" + configurationPort +"/config";
            var task = Task.Run(() => client.GetAsync(rqst));
            task.Wait();
            var rsp = task.Result;
            if (rsp.IsSuccessStatusCode == false)
            {
                throw new System.InvalidOperationException("Fail to get client configuration " + rqst);
            }

            // Here I will parse the JSON object and initialyse values from it...
            var serverConfig = JsonSerializer.Deserialize<ServerConfig>(rsp.Content.ReadAsStringAsync().Result);
            ServiceConfig config = null;
            if (!serverConfig.Services.ContainsKey(id))
            {
                foreach(var s in serverConfig.Services.Values)
                {
                    if(s.Name == id){
                        config = s;
                        break;
                    }
                }
                if(config == null){
                    throw new System.InvalidOperationException("No serivce found with id " + id + "!");
                }
            }else{
                config = serverConfig.Services[id];
            }

            // get the service config.
            this.port = config.Port;
            this.hasTls = config.TLS;
            this.domain = config.Domain;
            this.id = config.Id;
            this.name = config.Name;

            // Here I will create grpc connection with the service...
            if (!this.HasTLS())
            {
                // Non secure connection.
                this.channel = new Channel(this.domain, this.port, ChannelCredentials.Insecure);
            }
            else
            {
                // if the client is not local I will generate TLS certificates.
                if (File.Exists(Path.GetTempPath() + "/" + this.domain + "_token"))
                {
                    // The ca certificate.
                    this.caFile = config.CertAuthorityTrust;

                    // get the client certificate and key here.
                    this.certFile = config.CertFile.Replace("server", "client");
                    this.keyFile = config.KeyFile.Replace("server", "client");
                }
                else
                {
                    // I will need to create certificate and make it sign by the CA.
                    var path = Path.GetTempPath() + "/config/tls/" + this.domain;

                    if (!Directory.Exists(path))
                    {
                        Directory.CreateDirectory(path);
                    }

                    // Now I will create the certificates.
                    var ca_crt = getCaCertificate(this.domain,  configurationPort);
                    File.WriteAllText(path + "/ca.crt", ca_crt);

                    var pwd = "1111"; // Set in the configuration...

                    // Now I will generate the certificate for the client...
                    // Step 1: Generate client private key.
                    this.generateClientPrivateKey(path, pwd);

                    // Step 2: Generate the client signing request.
                    this.generateClientCertificateSigningRequest(path, this.domain);

                    // Step 3: Generate client signed certificate.
                    var client_csr = File.ReadAllText(path + "/client.csr");
                    var client_crt = this.signCaCertificate(this.domain, configurationPort, client_csr);
                    File.WriteAllText(path + "/client.crt", client_crt);

                    // Step 4: Convert client.key to pem file.
                    this.keyToPem("client", path, pwd);

                    // Set path in the config.
                    this.keyFile = path + "/client.key";
                    this.caFile = path + "/ca.crt";
                    this.certFile = path + "/client.crt";
                }

                var cacert = File.ReadAllText(this.caFile);
                var clientcert = File.ReadAllText(this.certFile);
                var clientkey = File.ReadAllText(this.keyFile);
                var ssl = new SslCredentials(cacert, new KeyCertificatePair(clientcert, clientkey), VerifyPeer);

                //File.WriteAllText("c:/temp/toto.txt", clientkey);
                this.channel = new Channel(this.domain, this.port, ssl);
            }
        }

        protected Metadata GetClientContext(string token = "", string application = "", string domain = "", string path = "")
        {
            // Set the token in the metadata.
            var metadata = new Metadata();

            // Here I will get the token from the file.
            if (token.Length == 0)
            {
                var path_ = Path.GetTempPath() + "/" + this.domain + "_token";
                if (File.Exists(path_))
                {
                    token = File.ReadAllText(path_);
                    metadata.Add("token", token);
                }
            }
            else
            {
                metadata.Add("token", token);
            }

            // set the local domain.
            if (domain.Length == 0)
            {
                metadata.Add("domain", this.domain);
            }
            else
            {
                metadata.Add("domain", domain);
            }

            if (application.Length > 0)
            {
                metadata.Add("application", application);
            }

            // The path of ressource if there one.
            if (path.Length > 0)
            {
                metadata.Add("path", path);
            }

            return metadata;
        }

        public Client(string id, string domain, int configurationPort)
        {
            this.id = id;
            this.domain = domain;
        
            // Now I will get the client configuration.
            this.init(id, domain, configurationPort);
        }
    }
}

File: services/cpp/GlobularClient/globularclient.cpp
#include "globularclient.h"
#include "HTTPRequest.hpp"
#include <sstream>
#include <fstream>
#include <iostream>
#include <cstdio>
#include <memory>
#include <stdexcept>
#include <string>
#include <array>
#include <filesystem>

//  https://github.com/nlohmann/json
#include "json.hpp"
#include "Base64.h"
#include <grpc/grpc.h>
#include <grpcpp/channel.h>
#include <grpcpp/create_channel.h>
#include <grpcpp/security/credentials.h>


/**
 * @brief writeAllText Write a text file into a given path.
 * @param path The path where to write the file.
 * @param text The text to save.
 */
void writeAllText(std::string path, std::string text){
    std::ofstream f;
    f.open(path);
    f << text;
    f.close();
}

/**
 * @brief readAllText Read all text from a given file
 * @param path The path of the file to be read.
 * @return
 */
std::string readAllText(std::string path){
    std::ifstream t(path);
    std::string str;

    t.seekg(0, std::ios::end);
    str.reserve(t.tellg());
    t.seekg(0, std::ios::beg);

    str.assign((std::istreambuf_iterator<char>(t)),
               std::istreambuf_iterator<char>());

    return str;
}

std::string getTempDir(){
    // std::filesystem::temp_directory_path()
    return std::filesystem::temp_directory_path().string();
}

bool exists(std::string path){
    //
    std::ifstream infile(path);
    return infile.good();
}

void createDir(std::string path){
    std::filesystem::create_directory(path);
}

std::string getPathName(const std::string& s) {

    char sep = '/';

#ifdef _WIN32
    sep = '\\';
#endif

    size_t i = s.rfind(sep, s.length());
    if (i != std::string::npos) {
        return(s.substr(0, i));
    }

    return("");
}
Globular::Client::Client(std::string name, std::string domain, unsigned int configurationPort)
{
    this->config = new Globular::ServiceConfig();
    this->config->Name = name;
    this->config->Domain = domain;

    // init internal values.
    this->init(configurationPort);
    std::cout << "Domain config port " << configurationPort << std::endl;
    std::stringstream ss;
    ss << this->config->Domain << ":" << this->config->Port;

    // Now I will create the grpc channel.
    if(this->getTls()){
        grpc::SslCredentialsOptions opts =
        {
            readAllText( this->config->CertAuthorityTrust),
            readAllText( this->config->KeyFile),
            readAllText( this->config->CertFile)
        };
        std::cout << "try to open channel with address " << ss.str() << std::endl;

        this->channel = grpc::CreateChannel(ss.str(), grpc::SslCredentials ( opts ) );

    }else{
        // Create insecure connection to the service.
        this->channel = grpc::CreateChannel(ss.str(), grpc::InsecureChannelCredentials());
        if(this->channel){
            std::cout << "client channel is now initialysed!" << std::endl;
            std::cout << this->channel->GetServiceConfigJSON()	<< std::endl;
        }
    }

}

void Globular::Client::init(unsigned int configurationPort){

    // Initialyse client stuff here.
    this->initServiceConfig(configurationPort);

}

void Globular::Client::initServiceConfig(unsigned int configurationPort){

    std::stringstream ss;
    ss << "http://" << this->config->Domain << ":" << configurationPort << "/config";
    std::cout << "get configuration from " << ss.str() << std::endl;;
    http::Request request(ss.str());
    const http::Response response = request.send("GET");
    ss.flush();
    ss << std::string(response.body.begin(), response.body.end()) << '\n'; // print the result
    std::string jsonStr = ss.str();
    size_t index = jsonStr.find_first_of("{");
    jsonStr = jsonStr.substr(index, jsonStr.length() - index);
    auto j = nlohmann::json::parse(jsonStr);

    // Now I will initialyse the value from the configuration file.
    auto services = j["Services"];

    //
    for (auto it = services.begin(); it != services.end(); ++it)
    {
        if(it.key() == this->config->Name || (*it)["Id"].get<std::string>() == this->config->Name){

            this->config->Id = (*it)["Id"].get<std::string>();
            this->config->Name = (*it)["Name"].get<std::string>();
            this->config->Domain = (*it)["Domain"].get<std::string>();
            this->config->Description = (*it)["Domain"].get<std::string>();
            this->config->Port = (*it)["Port"].get<unsigned int>();
            this->config->Proxy = (*it)["Proxy"].get<unsigned int>();
            this->config->TLS= (*it)["TLS"].get<bool>();
            this->config->Keywords = (*it)["Keywords"].get<std::vector<std::string>>();
            this->config->Discoveries = (*it)["Discoveries"].get<std::vector<std::string>>();
            this->config->Repositories = (*it)["Repositories"].get<std::vector<std::string>>();

            if(this->config->TLS){
                // The service is secure.
                std::stringstream ss;
                ss << getTempDir() <<   "/" << this->config->Domain + "_token";
                auto path = ss.str();
                std::cout << "read: " << path << std::endl;
                if(exists(path)){
                    // TODO make correction here the CertFile and KeyFile are the one of the server not the client.
                    this->config->CertAuthorityTrust =  (*it)["CertAuthorityTrust"].get<std::string>();
                    std::string path = getPathName(this->config->CertAuthorityTrust);

                    this->config->CertFile =  path + "/client.crt";
                    this->config->KeyFile =   path + "/client.pem";;

                    std::cout << this->config->KeyFile << std::endl;
                    std::cout << this->config->CertFile << std::endl;
                    std::cout << this->config->CertAuthorityTrust << std::endl;
                }else{
                    ss.flush();
                    ss << getTempDir() <<   "/config/tls/" << this->config->Domain;
                    auto path = ss.str();
                    // Here I will create a directory named
                    if(!exists(path)){
                        createDir(path);
                    }

                    // I will create a certificate request and make it sign by the server.
                    auto ca_crt = this->getCaCertificate(this->config->Domain, configurationPort);
                    writeAllText(path + "/ca.crt", ca_crt);

                    // The password must be store in the client configuration...
                    auto pwd = "1111";

                    // Now I will generate the certificate for the client...
                    // Step 1: Generate client private key.
                    this->generateClientPrivateKey(path, pwd);

                    // Step 2: Generate the client signing request.
                    this->generateClientCertificateSigningRequest(path, this->config->Domain);


                    // Step 3: Generate client signed certificate.
                    auto client_csr = readAllText(path + "/client.csr");
                    auto client_crt = this->signCaCertificate(this->config->Domain, configurationPort, client_csr);
                    writeAllText(path + "/client.crt", client_crt);

                    // Step 4: Convert client.key to pem file.
                    this->keyToPem("client", path, pwd);

                    // Set path in the config.
                    this->config->KeyFile= path + "/client.key";
                    this->config->CertAuthorityTrust  = path + "/ca.crt";
                    this->config->CertFile  = path + "/client.crt";
                }
            }
            break; // exit the loop
        }
    }
}

std::string Globular::Client::getName() const
{
    return this->config->Name;
}

std::string Globular::Client::getDomain() const
{
    return this->config->Domain;
}

unsigned int Globular::Client::getPort() const
{
    return this->config->Port;
}

bool Globular::Client::getTls() const
{
    return this->config->TLS;
}

std::string Globular::Client::getCaFile() const
{
    return this->config->CertAuthorityTrust;
}

void Globular::Client::setCaFile(const std::string &value)
{
    this->config->CertAuthorityTrust = value;
}

std::string Globular::Client::getKeyFile() const
{
    return this->config->KeyFile;
}

void Globular::Client::setKeyFile(const std::string &value)
{
    this->config->KeyFile = value;
}

std::string Globular::Client::getCertFile() const
{
    return this->config->CertFile;
}

void Globular::Client::setCertFile(const std::string &value)
{
    this->config->CertFile = value;
}

void Globular::Client::setTls(bool value)
{
    this->config->TLS = value;
}

void Globular::Client::close()
{
}

void Globular::Client::getClientContext(ClientContext& context, std::string token, std::string application, std::string domain, std::string path){
    if(domain.empty()){
        context.AddMetadata("domain", this->config->Domain);
        domain = this->config->Domain;
    }else{
        context.AddMetadata("domain", domain);
    }

    if(token.empty()){
        std::stringstream ss;
        ss << getTempDir() << "/" << domain << "_token";
        if(exists(ss.str())){
            token = readAllText(ss.str());
            context.AddMetadata("token", token);
        }
    }else{
        context.AddMetadata("token", token);
    }

    if(!application.empty()){
        context.AddMetadata("application", application);
    }

    if(!path.empty()){
        context.AddMetadata("path", path);
    }
}

std::string exec(const char* cmd) {
    std::array<char, 128> buffer;
    std::string result;
    std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd, "r"), pclose);
    if (!pipe) {
        throw std::runtime_error("popen() failed!");
    }
    while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {
        result += buffer.data();
    }
    return result;
}

std::string Globular::Client::getCaCertificate(std::string domain, unsigned int configurationPort){
    std::stringstream ss;
    ss << "http://" << domain << ":" << configurationPort << "/get_ca_certificate";
    http::Request request(ss.str());
    const http::Response response = request.send("GET");
    ss.flush();
    ss << std::string(response.body.begin(), response.body.end()) << '\n'; // print the result
    return ss.str();
}

std::string Globular::Client::signCaCertificate(std::string domain, unsigned int configurationPort, std::string csr){
    std::stringstream ss;
    std::string csr_str = macaron::Base64::Encode(csr);
    ss << "http://" << domain << ":" << configurationPort << "/sign_ca_certificate?=" + csr_str;
    http::Request request(ss.str());
    const http::Response response = request.send("GET");
    ss.flush();
    ss << std::string(response.body.begin(), response.body.end()) << '\n'; // print the result
    return ss.str();
}

// TODO fix to new certificate with multiple domains.
void Globular::Client::generateClientPrivateKey(std::string path, std::string pwd){
    std::stringstream ss;
    ss << path <<   "/client.key";
    auto path_ = ss.str();
    if(exists(path)){
        return;
    }

    ss.flush();

    ss << "openssl.exe";
    ss<< " genrsa";
    ss << " -passout";
    ss << " pass:"  << pwd;
    ss << " -des3";
    ss << " -out ";
    ss << " " << path << "/client.pass.key";
    ss << " 4096";
    std::system(ss.str().c_str());

    ss.flush();
    ss << "openssl.exe";
    ss<< " genrsa";
    ss << " -passin";
    ss << " pass:"  << pwd;
    ss << " -in";
    ss << " " << path << "/client.pass.key";
    ss << " -out ";
    ss << " " << path << "/client.key";
    std::system(ss.str().c_str());

    ss.flush();
    ss << path << "/client.pass.key";
    remove(ss.str().c_str());

}

void Globular::Client::generateClientCertificateSigningRequest(std::string path, std::string domain){
    std::stringstream ss;
    ss << path <<   "/client.csr";
    auto path_ = ss.str();
    if(exists(path)){
        return;
    }

    ss.flush();

    ss << "openssl.exe";
    ss<< " req";
    ss << " -new";
    ss << " -key";
    ss << " " << path << "/client.key";
    ss << " -out ";
    ss << " " << path << "/client.csr";
    ss << " -subj";
    ss << " /CN=" << domain;

    // run the command...
    std::system(ss.str().c_str());
}

void Globular::Client::keyToPem(std::string name, std::string path, std::string pwd){
    std::stringstream ss;
    ss << path <<   "/" << name + ".pem";
    auto path_ = ss.str();
    if(exists(path)){
        return;
    }

    ss.flush();

    ss << "openssl.exe";
    ss<< " pkcs8";
    ss << " -topk8";
    ss << " -nocrypt";
    ss << " -passin";
    ss << " pass:"  << pwd;
    ss << " -in ";
    ss << " " << path << "/" << name << ".key";
    ss << " -out ";
    ss << " " << path << "/" << name << ".pem";

    // run the command...
    std::system(ss.str().c_str());
}
